### PR TITLE
SAGE 1249: Add vsn

### PR DESCRIPTION
**Pull request by iperezx**
Summary of changes:
- Allow for quiet mode in the node_ssh function. This allows to be able to just directly get data from the command that is sent through the ssh tunnerl. Defaults are left as before.
- I am just doing a simple `cat /etc/waggle/vsn` on the node.
- Add new unit test for vsn.

### CODE REVIEW

**Comment by iperezx** [comment]
Takeaways from demo:
-  Move "get vsn" to the registration process

**Comment by iperezx** [comment]
@jswantek @seanshahkarami @nconrad 

What do you guys think?
Currently, the trigger to get the vsn is using this curl command:
```
curl localhost:5000/node/0000000000000001 -d '{"vsn": true}'
```

I vaguely remember that you guys didnt like this way of interacting with the API. From my part, I was just mimicking how things are done in the API and it was simple. All the tools to trigger beekeeper to get the VSN through the SSH tunnel is completed and with unit test. Now it's just a matter of how we want to interact with the API. 

Based on my JIRA ticket this is the last comment I wrote and what caused our tangent to the node-manifest discussions:
```
There are two type of actions that we need with VSN:
The first one is to check that VSN is present in the node -> check that vsn is null and grab the vsn on the node to update the null value
The second one is to check that the VSN is up-to-date. -> I get the current VSN in the db and check against the nodes vsn.
```

File: bk-api/test_bk_api.py [hunk 1]
```diff
@@ -139,6 +139,30 @@ def test_log_insert_fail(client):
...
+    assert "success" in result
+    #Check that the db is updated with the new value
+    gt_value = "TEST-minimal"
+    rv = client.get(f'/state/0000000000000001')
+    result = rv.get_json()
+    print(result)
+    assert 'error' not in result
+    assert 'data' in result
+
+    d_after_vsn = result["data"]["vsn"]
```
Comment (jswantek) [inline bk-api/test_bk_api.py:163]: can u move this to right after the `rv.get_json()` call like you did on line 146?

its easier to compare the "before" and "after"

File: bk-api/test_bk_api.py [hunk 2]
```diff
@@ -139,6 +139,30 @@ def test_log_insert_fail(client):
...
+    assert 'error' not in result_before_vsn
+    assert 'data' in result_before_vsn
+    assert None == d_before_vsn
+    #Check that we can populate vsn
+    data = {"vsn": True}
+    rv = client.post('/node/0000000000000001', data = json.dumps(data))
+    result = rv.get_json()
+    assert "success" in result
+    #Check that the db is updated with the new value
+    gt_value = "TEST-minimal"
```
Comment (jswantek) [inline bk-api/test_bk_api.py:156]: why "TEST-minimal" ?  where does this come from?  how does node `0000000000000001` have a VSN of "TEST-minimal"?  Where did you define that on your test node?

File: bk-api/bk_api.py [hunk 3]
```diff
@@ -585,17 +585,25 @@ def scp(source, node_id, target):
 
 
 
-def node_ssh(node_id, command, input_str=None):
+def node_ssh(node_id, command, input_str=None,quiet_mode=False):
```
Comment (jswantek) [inline bk-api/bk_api.py:588]: please run "linting" on your changes  (and maybe the whole file) to ensure normalization of white spaces and tabs.

i.e. there is no space here: `input_str=None,quiet_mode=False`

**Comment by jswantek** [comment]
Please use JIRA ticket numbers in your git commit comments

```
SAGE-1234: <short description>

<optional longer description>
```

File: bk-api/bk_api.py [hunk 4]
```diff
@@ -585,17 +585,25 @@ def scp(source, node_id, target):
...
+    proxy_cmd = f"ProxyCommand=ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@{BEEKEEPER_SSHD_HOST} -p 2201 -i /config/admin-key/admin.pem"
```
Comment (jswantek) [inline bk-api/bk_api.py:589]: why dont you put the `-o` in the `proxy_cmd`

```
proxy_cmd = ["-o" , f"ProxyCommand=ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@{BEEKEEPER_SSHD_HOST} -p 2201 -i /config/admin-key/admin.pem"]
```

File: bk-api/bk_api.py [hunk 5]
```diff
@@ -585,17 +585,25 @@ def scp(source, node_id, target):
...
-            #"-tt",
-            "-i", node_key , # this must be the key that allows ssh into the node , this key might be the same for all nodes
-            "-o", "UserKnownHostsFile=/dev/null",
-            "-o", "StrictHostKeyChecking=no",
-            "-o", "IdentitiesOnly=true",
-            "-o" ,f"ProxyCommand=ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@{BEEKEEPER_SSHD_HOST} -p 2201 -i /config/admin-key/admin.pem  netcat -U /home_dirs/node-{node_id}/rtun.sock",
-            f"root@foo:",
-            command]
-
+                #"-tt",
```
Comment (jswantek) [inline bk-api/bk_api.py:591]: 1) this change seems unrelated to your VSN change and therefore be it's own commit in the event we need to revert either the VSN change or this ssh change

2) I would have done this a little different (for reference)

```
proxy_sub_cmd = f"ProxyCommand=ssh -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no root@{BEEKEEPER_SSHD_HOST} -p 2201 -i /config/admin-key/admin.pem"
if quiet_mode:
    proxy_sub_cmd += " -q"
proxy_sub_cmd += f" netcat -U /home_dirs/node-{node_id}/rtun.sock"

proxy_cmd = ["-o", f"{proxy_sub_cmd}"]

ssh_options = proxy_cmd + [
    "-o",
    "UserKnownHostsFile=/dev/null",
    "-o",
    "StrictHostKeyChecking=no",
    "-o",
    "IdentitiesOnly=true",
]

ssh_cmd = [
    "ssh",
    "-i",
    node_key,  # this must be the key that allows ssh into the node , this key might be the same for all nodes
] + proxy_cmd
if quiet_mode:
    ssh_cmd.append("-q")
ssh_cmd.append(f"root@foo:")
ssh_cmd.append(command)
```

File: bk-api/bk_api.py [hunk 6]
```diff
@@ -1077,6 +1117,15 @@ def post(self, node_id):
 
             return jsonify(result)
 
+        if "vsn" in postData:
```
Comment (jswantek) [inline bk-api/bk_api.py:1120]: i think we want to do this before `deploy_wes` , how do ensure that?

File: bk-api/bk_api.py [hunk 7]
```diff
@@ -903,7 +911,39 @@ def deploy_wes(node_id, this_debug, force=False):
...
+    try:
+        insert_log(payload, lock_tables=lock_tables, lock_requested_by=lock_requested_by)
+    except Exception as e:
+        raise Exception(f"insert_log returned: {str(e)}")
+    return
+
+def add_vsn(node_id):
+    command = "cat /etc/waggle/vsn"
+    try:
+        result_stdout_str ,result_stderr_str, exit_code = node_ssh(node_id, command,quiet_mode=True)
```
Comment (jswantek) [inline bk-api/bk_api.py:929]: `,` and `<space>` yucko.. please normalize

File: bk-api/bk_api.py [hunk 8]
```diff
@@ -903,7 +911,39 @@ def deploy_wes(node_id, this_debug, force=False):
...
+    vsn_val_str = str(result_stdout_str).rstrip()
+    #is vsn empty or not an instance of str?
+    if isinstance( vsn_val_str, str ):
+        if vsn_val_str.strip() == '':
+            raise Exception(f"vsn is empty: {vsn_val_str}")
+    else:
+        raise Exception(f"vsn invalid type")
 
+    try:
+        register_add_vsn_event(node_id,vsn_val_str,lock_tables=True, lock_requested_by="vsn")
```
Comment (jswantek) [inline bk-api/bk_api.py:943]: can you be more specific for `lock_requested_by` .. `"vsn_retrieval"`

File: bk-api/bk_api.py [hunk 9]
```diff
@@ -903,7 +911,39 @@ def deploy_wes(node_id, this_debug, force=False):
     return {"success":True}
 
 
+def register_add_vsn_event(node_id,field_value, lock_tables=True, lock_requested_by=""):
+
+    payload = {"node_id": node_id, "source": "beekeeper-add-vsn", "operation":"insert", 
+                "field_name": "vsn", "field_value": field_value}
+
+    #url = f'{BEEKEEPER_DB_API}/log'
```
Comment (jswantek) [inline bk-api/bk_api.py:919]: remove

**Review by jswantek — CHANGES_REQUESTED** [review]

File: schema.sql [hunk 10]
```diff
@@ -26,6 +26,7 @@ CREATE TABLE IF NOT EXISTS Beekeeper.nodes_log (
 CREATE TABLE IF NOT EXISTS Beekeeper.nodes_history (
     `id`                    VARCHAR(16) NOT NULL,  /*  typically MAC address of the "main" device  */
     `timestamp`             TIMESTAMP(0) NOT NULL,
+    `vsn`                   VARCHAR(64),
```
[the remark this answers is not in the document]
Comment (iperezx) [reply schema.sql:29]: 1. Good question. I believe the same way you guys added the `wes_deploy_event` column.
2. Similar to the deploy_wes service. Just fetch it from the node.

File: README.md [hunk 11]
```diff
@@ -51,6 +51,7 @@ This will only work once node has registered.
 ```bash
 curl localhost:5000/node/0000000000000001 -d '{"assign_beehive": "my-beehive"}'
 curl localhost:5000/node/0000000000000001 -d '{"deploy_wes": true}'
+curl localhost:5000/node/0000000000000001 -d '{"vsn": true}'
```
[the remark this answers is not in the document]
Comment (iperezx) [reply README.md:54]: this is not an ordered list and we can get the VSNs first. There are no dependencies on either one at this point.

File: bk-api/test_bk_api.py (hunk 1 above)
Comment (iperezx) [reply bk-api/test_bk_api.py:163]: That was my accident. Made the change just now. 
I need the assert statements before I pass in a key into the dictionary. If not I will error out before I get to the assert.

File: bk-api/bk_api.py (hunk 3 above)
Comment (iperezx) [reply bk-api/bk_api.py:588]: If it is just a whitespace, I will fix it but turning on pylint will complain about everything. Are you guys using a specific config file for pylint?

File: bk-api/bk_api.py (hunk 4 above)
Comment (iperezx) [reply bk-api/bk_api.py:589]: It's easier/simpler to just append to a string when I add `-q` to the proxy_cmd.

File: bk-api/bk_api.py (hunk 5 above)
Comment (iperezx) [reply bk-api/bk_api.py:591]: 1. The change is needed so that the cat command (https://github.com/iperezx/beekeeper/blob/master/bk-api/bk_api.py#L927) only contains the VSN value. If not you get a lot of output in stdout from the ssh connection. 
2. Isn't that just style? I wanted to keep it in one area instead of just scattered around the functon.

File: bk-api/bk_api.py (hunk 9 above)
Comment (iperezx) [reply bk-api/bk_api.py:919]: Why?

File: bk-api/bk_api.py (hunk 6 above)
Comment (iperezx) [reply bk-api/bk_api.py:1120]: Ensure what? It just needs to call that endpoint and it will get the VSN from the node. Very similar to the deploy_wes service.

File: schema.sql (hunk 10 above)
[the remark this answers is not in the document]
Comment (jswantek) [reply schema.sql:29]: 1. can you investigate if we need versions in our tables and a "upgrade" plan.  i thought i remember (in past projects) there being code that could run when upgrading a table
2. yeah we can just do it in post for every node.

File: README.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (jswantek) [reply README.md:54]: please update the documentation to suggest the proper order

File: bk-api/bk_api.py (hunk 5 above)
Comment (jswantek) [reply bk-api/bk_api.py:591]: 1. i would still do the ssh change as a separate commit... or are you planning on merge squashing this?  i would prefer we did not do that.  i like to have separate well defined commits that you can reference back to and revert if needed.  this is why i do rebase on my PRs
2. see.. from my perspective the configuration of the proxy was already scattered around the function. ;)

File: bk-api/bk_api.py (hunk 9 above)
Comment (jswantek) [reply bk-api/bk_api.py:919]: you need the commented out line of code?

```
#url = f'{BEEKEEPER_DB_API}/log'
```

File: bk-api/bk_api.py (hunk 6 above)
Comment (jswantek) [reply bk-api/bk_api.py:1120]: when a node registers with beekeeper the steps that are taken are 
- `deploy_wes`
- `vsn`
how do you ensure the `vsn` step is done before `deploy_wes` on node registration?

File: schema.sql (hunk 10 above)
[the remark this answers is not in the document]
Comment (iperezx) [reply schema.sql:29]: 1. When this commit was made: https://github.com/waggle-sensor/beekeeper/commit/5497810a6a57fd04f59653012aea99af00138efc#diff-939acd018b931292330e519becd8d003ed0dd20c5e42de62ec3d37553698739a , beekeeper was already deployed, correct?

File: bk-api/bk_api.py (hunk 6 above)
Comment (iperezx) [reply bk-api/bk_api.py:1120]: That is done externally by the bk-deploy-manager for deploy_wes, but VSN is not added there yet. 
Not sure where you guys want to add it because we need to first answer these questions (my comment from yesterday):
```
There are two type of actions that we need with VSN:
The first one is to check that VSN is present in the node -> check that vsn is null and grab the vsn on the node to update the null value
The second one is to check that the VSN is up-to-date. -> I get the current VSN in the db and check against the nodes vsn.
```

File: bk-api/bk_api.py (hunk 5 above)
Comment (iperezx) [reply bk-api/bk_api.py:591]: 1. I have no preference. I can separate it.
2.😃

File: schema.sql (hunk 10 above)
[the remark this answers is not in the document]
Comment (jswantek) [reply schema.sql:29]: I have no idea.  but we should establish a process for how we will be upgrading our database(s) in production?  to ensure we dont blow away something.

ref: https://dev.mysql.com/doc/refman/5.7/en/mysql-upgrade.html

File: bk-api/bk_api.py (hunk 3 above)
Comment (jswantek) [reply bk-api/bk_api.py]: i use the VSCode linter: https://marketplace.visualstudio.com/items?itemName=ms-python.python

On save it formats the entire file.  What i tend to do is make 1 commit that is an auto-lint of the entire file. then i make my changes on top of that.

Comment (jswantek) [reply bk-api/bk_api.py]: Its really your call here.. but if we all use the same linting techniques our code will be ultra-pretty and look the same.

File: bk-api/bk_api.py (hunk 6 above)
Comment (jswantek) [reply bk-api/bk_api.py:1119]: Lets talk about this as a team.  as I don't remember why we wouldn't just do this on Node registration.  as the VSN doesn't change very often at all.  and any manual changes of VSN (which I only do very very rarely) would require me to just do a new `curl localhost:5000/node/0000000000000001 -d '{"vsn": true}'`

File: bk-api/bk_api.py (hunk 3 above)
Comment (iperezx) [reply bk-api/bk_api.py]: Sounds good. The other way we can do it is in the CI to run a linter with a specific style guide so anything that is added to main/master will be abiding a specific style guide.

File: bk-api/test_bk_api.py (hunk 2 above)
Comment (seanshahkarami) [reply bk-api/test_bk_api.py:156]: I agree totally agree.

At some point, we started putting some Vagrant stuff there to help run it, so the repo became less "pure" k8s stuff.

I'd prefer to split.

File: schema.sql (hunk 10 above)
[the remark this answers is not in the document]
Comment (seanshahkarami) [reply schema.sql:29]: This would fall under "migrations". In this case, we'd need to do an ALTER TABLE add this field.

Unfortunately, this is generally a challenging thing, especially when all the DB abstraction is written by hand...

ORMs do a lot of work to make this possible. For example, in frameworks like Django, as you update your database models, it generates the steps to migrate the tables safely and covers stuff like adding new columns, prompting you to backfill a new required column, preventing bad types, etc.

All that's to say, it's a nontrivial thing we need to think about.

[the remark this answers is not in the document]
Comment (seanshahkarami) [reply schema.sql:29]: This would fall under "migrations". In this case, we'd need to do an ALTER TABLE add this field.

Unfortunately, this is generally a challenging thing, especially when all the DB abstraction is written by hand...

ORMs do a lot of work to make this possible. For example, in frameworks like Django, as you update your database models, it generates the steps to migrate the tables safely and covers stuff like adding new columns, prompting you to backfill a new required column, preventing bad types, etc.

All that's to say, it's a nontrivial thing we need to think about.

[the remark this answers is not in the document]
Comment (seanshahkarami) [reply schema.sql:29]: To be totally honest... part of me wonders if we should consider porting some of this layer to a framework like Django to help abstract details like this automatically.

We can maintain all the APIs as they are now, but it may pay off in the long run.

File: bk-api/bk_api.py (hunk 6 above)
Comment (seanshahkarami) [reply bk-api/bk_api.py:1119]: I agree. I think we should make this a registration time thing.

My own preference is generally to not "reach into the node", if possible. That would also allow more generic testing of registration, since a test program can provide all that info, even when there's no real node backing it.

Comment (seanshahkarami) [reply bk-api/bk_api.py:1119]: I agree. I think we should make this a registration time thing.

My own preference is generally to not "reach into the node", if possible. That would also allow more generic testing of registration, since a test program can provide all that info, even when there's no real node backing it.

File: schema.sql (hunk 10 above)
[the remark this answers is not in the document]
Comment (iperezx) [reply schema.sql:29]: ORM will be ideal for this scenario. I can put together something to just show the data model.

I will prefer to use FastAPI(https://fastapi.tiangolo.com/) instead of django.

[the remark this answers is not in the document]
Comment (seanshahkarami) [reply schema.sql:29]: You mean something like sqlalchemy? (Since I don't think FastAPI does database / model management on its own?)

It's your call. I really love FastAPI... but for a (potentially) very CRUD type app, it seems like Django + Django REST Framework _might_ be a better fit.

Especially with things like Django's builtin admin UI.

[the remark this answers is not in the document]
Comment (iperezx) [reply schema.sql:29]: It uses sqlalchemy in the backend for ORM in fastapi: https://fastapi.tiangolo.com/tutorial/sql-databases/#orms

[the remark this answers is not in the document]
Comment (jswantek) [reply schema.sql:29]: Taken note of this for future work

File: bk-api/test_bk_api.py [hunk 12]
```diff
@@ -139,6 +139,30 @@ def test_log_insert_fail(client):
     assert 'error'  in result
 
 
+def test_vsn_insert(client):
```
Comment (seanshahkarami) [inline bk-api/test_bk_api.py:142]: I think we should have a test for `register_add_vsn_event(node_id, vsn)` to make sure you can set a a few VSNs and read them back out?

This is more like integration testing and I'm not following exactly when and how these tests are run.

Maybe we can start working towards better defining the interfaces around the:
* Database
* API
* Processes like "deploy WES" and "fetch VSN"

With clean boundaries, it will be easy for us to just write a unit test like this for just the database portion.

File: bk-api/bk_api.py [hunk 13]
```diff
@@ -903,7 +911,38 @@ def deploy_wes(node_id, this_debug, force=False):
...
+    vsn_val_str = str(result_stdout_str).rstrip()
+    #is vsn empty or not an instance of str?
+    if isinstance( vsn_val_str, str ):
+        if vsn_val_str.strip() == '':
+            raise Exception(f"vsn is empty: {vsn_val_str}")
+    else:
+        raise Exception(f"vsn invalid type")
+
+    try:
+        register_add_vsn_event(node_id,vsn_val_str,lock_tables=True, lock_requested_by="vsn_retrieval")
```
[the remark this answers is not in the document]
Comment (seanshahkarami) [reply bk-api/bk_api.py:942]: I'm only saying this as it clutters the code and this probably isn't the only place it appears...

**Comment by jswantek** [comment]
You will need to rebase this as the `origin` branch changed from `master` to `main`
